mintscript/src/translate.rs

```diff
@@ -0,0 +1,231 @@
+
+    #[error(transparent)]
+    PoSAccounting(#[from] pos_accounting::Error),
+
+    #[error(transparent)]
+    TokensAccounting(#[from] tokens_accounting::Error),
+
+    #[error("Stake pool {0} does not exist")]
+    PoolNotFound(PoolId),
+
+    #[error("Delegation {0} does not exist")]
+    DelegationNotFound(DelegationId),
+
+    #[error("Token with id {0} does not exist")]
+    TokenNotFound(TokenId),
+}
+
+/// Contextual information about given input
+pub enum InputInfo<'a> {
+    Utxo {
+        outpoint: &'a UtxoOutPoint,
+        utxo: Utxo,
+    },
+    Account {
+        outpoint: &'a AccountOutPoint,
+    },
+    AccountCommand {
+        command: &'a AccountCommand,
+    },
+}
+
+impl InputInfo<'_> {
+    pub fn as_utxo_output(&self) -> Option<&TxOutput> {
+        match self {
+            InputInfo::Utxo { outpoint: _, utxo } => Some(utxo.output()),
+            InputInfo::Account { .. } | InputInfo::AccountCommand { .. } => None,
+        }
+    }
+}
+
+pub trait InputInfoProvider {
+    fn input_info(&self) -> &InputInfo;
+    fn witness(&self) -> &InputWitness;
+}
+
+/// Provides information necessary to translate an input to a script.
+pub trait SignatureInfoProvider: InputInfoProvider {
+    type PoSAccounting: PoSAccountingView<Error = pos_accounting::Error>;
+    type Tokens: TokensAccountingView<Error = tokens_accounting::Error>;
+
+    fn pos_accounting(&self) -> &Self::PoSAccounting;
+    fn tokens(&self) -> &Self::Tokens;
+}
+
+pub trait TranslateInput<C> {
+    /// Translate transaction input to script.
+    fn translate_input(ctx: &C) -> Result<WitnessScript, TranslationError>;
+}
+
+impl<C: SignatureInfoProvider> TranslateInput<C> for SignedTransaction {
+    fn translate_input(ctx: &C) -> Result<WitnessScript, TranslationError> {
+        let pos_accounting = ctx.pos_accounting();
+        let checksig =
+            |dest: &Destination| WitnessScript::signature(dest.clone(), ctx.witness().clone());
+
+        match ctx.input_info() {
+            InputInfo::Utxo { outpoint: _, utxo } => match utxo.output() {
+                TxOutput::Transfer(_val, dest) => Ok(checksig(dest)),
+                TxOutput::LockThenTransfer(_val, dest, timelock) => {
+                    Ok(WitnessScript::satisfied_conjunction([
+                        WitnessScript::timelock(*timelock),
+                        checksig(dest),
+                    ]))
+                }
+                TxOutput::CreateStakePool(_pool_id, pool_data) => {
+                    Ok(checksig(pool_data.decommission_key()))
+                }
+                TxOutput::ProduceBlockFromStake(_dest, pool_id) => {
+                    let pool_data = pos_accounting
+                        .get_pool_data(*pool_id)?
+                        .ok_or(TranslationError::PoolNotFound(*pool_id))?;
+                    Ok(checksig(pool_data.decommission_destination()))
+                }
+                TxOutput::DelegateStaking(_amount, delegation_id) => {
```

Comment:
> Also I'd place a TODO to remove SignatureDestinationGetter

So far, it's a checkbox at the top of this PR. Will be turned into a TODO/issue or resolved.

Reply:
> this is not spendable. I see it's taken from SignatureDestinationGetter but that's actually a bug there (thoughn the code is effectively unreachable because no utxo is available).

Ok. Are you reasonably sure this will not cause forks? This is consensus code so I am a bit paranoid here.

Reply:
> I was hoping we could ditch SignatureDestinationGetter before merging this PR...

Tried to do it a while back but it's still used in tests to generate signatures. Can move it into a test module. In principle, it could be gutted out and the implementation replaced with something that translates the input to a script and extracts destinations.

Reply:
I am because such outputs are not included into utxo set see: https://github.com/mintlayer/mintlayer-core/blob/870434bd92f3c94d82f304efdc06ce0602a77d60/utxo/src/cache.rs#L495 so they cannot be spent and thus verified.

I don't see the test for that so you can add it with the patch:
[spend_delegate_staking_output_test.patch](https://github.com/user-attachments/files/15742718/spend_delegate_staking_output_test.patch)

Reply:
> Can move it into a test module

I'm all in for that.

Reply:
> > Can move it into a test module
> 
> I'm all in for that.

This is done

Reply:
> > Are you reasonably sure this will not cause forks?
> 
> I am because such outputs are not included into utxo set see:
> 
> https://github.com/mintlayer/mintlayer-core/blob/870434bd92f3c94d82f304efdc06ce0602a77d60/utxo/src/cache.rs#L495
> 
> so they cannot be spent and thus verified.
> I don't see the test for that so you can add it with the patch: [spend_delegate_staking_output_test.patch](https://github.com/user-attachments/files/15742718/spend_delegate_staking_output_test.patch)

Changed this and included the test. Thanks.